src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,240 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
             Common.ILogger log,
             CancellationToken token)
         {
-            var metadataCache = new MetadataReferenceCache();
-
-            var packages =
-                (await _regResource.GetPackageMetadata(packageId, includePrerelease, includeUnlisted, sourceCacheContext, log, token))
-                    .Select(ParseMetadata)
-                    .Select(m => metadataCache.GetObject(m))
-                    .ToArray();
+            var range = VersionRange.All; // This value preset for Consolidate UI.
 
-            return packages;
+            return await GetMetadata(packageId, includePrerelease, includeUnlisted, range, sourceCacheContext, log, token);
         }
 
+        /// <summary>
+        /// Returns the registration blob for the id and version
+        /// </summary>
+        /// <param name="package"></param>
+        /// <param name="sourceCacheContext"></param>
+        /// <param name="log"></param>
+        /// <param name="token"></param>
+        /// <returns>Package meta data.</returns>
+        /// <remarks>The inlined entries are potentially going away soon</remarks>
         public override async Task<IPackageSearchMetadata> GetMetadataAsync(
             PackageIdentity package,
             SourceCacheContext sourceCacheContext,
             Common.ILogger log,
             CancellationToken token)
         {
-            var metadata = await _regResource.GetPackageMetadata(package, sourceCacheContext, log, token);
-            if (metadata != null)
+            var range = new VersionRange(package.Version, true, package.Version, true);
+            var includePrerelease = true;
+            var includeUnlisted = true;
+
+            return (await GetMetadata(package.Id, includePrerelease, includeUnlisted, range, sourceCacheContext, log, token)).SingleOrDefault();
+        }
+
+        private async Task<IEnumerable<IPackageSearchMetadata>> GetMetadata(
+            string packageId,
+            bool includePrerelease,
+            bool includeUnlisted,
+            VersionRange range,
+            SourceCacheContext sourceCacheContext,
+            ILogger log,
+            CancellationToken token)
+        {
+            var metadataCache = new MetadataReferenceCache();
+            var registrationUri = _regResource.GetUri(packageId);
+
+            var registrationIndexResult = await LoadRangesAsync(
+                _client,
+                registrationUri,
+                packageId,
+                sourceCacheContext,
+                async httpSourceResult => await ProcessRegistrationIndexAsync<RegistrationIndex>(httpSourceResult, token),
```

Comment:
```suggestion
                httpSourceResult => ProcessRegistrationIndexAsync<RegistrationIndex>(httpSourceResult, token),
```

removes one async state machine and therefore memory allocation.

Reply:
Done.